### PR TITLE
feat(extension): support duplicate template

### PR DESCRIPTION
**Pull request by airslice**
## Overview

This PR adds the support for duplicate template
- Create a new template with the same settings
- Support component template and emphasis property template
- Automatically generate the name with timestamp, user can rename to what they want.

## Screenshot

image
image
image

## Summary by CodeRabbit

- **New Features**
	- Introduced a template duplication feature across multiple components, allowing users to easily create copies of existing templates.
	- Added a dialog for users to enter the name of the duplicated template, enhancing the template management experience.

- **Bug Fixes**
	- Improved error handling and user notifications during the template duplication process.

### CODE REVIEW

File: extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx [hunk 1]
```diff
@@ -138,6 +175,22 @@ export const TemplateHeader: React.FC<TemplateHeaderProps> = ({
...
+          onSubmit={handleTemplateDuplicate}
+          submitDisabled={duplicatedTemplateNameValid}>
+          <EditorTextField
+            autoFocus
+            label="Full path:"
+            fullWidth
+            value={duplicatedTemplateName}
+            onChange={handleDuplicatedTemplateNameChange}
+          />
+        </EditorDialog>
```
[the remark this answers is not in the document]
Comment (airslice) [reply extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx:193]: Updated.

File: extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx [hunk 2]
```diff
@@ -85,6 +87,38 @@
...
+    [],
+  );
+  const duplicatedTemplateNameValid = useMemo(() => {
+    return (
+      templateNames.includes(duplicatedTemplateName) ||
+      duplicatedTemplateName === "" ||
+      duplicatedTemplateName.endsWith("/") ||
+      duplicatedTemplateName.split("/").some(p => p === "")
+    );
+  }, [duplicatedTemplateName, templateNames]);
```
[the remark this answers is not in the document]
Comment (airslice) [reply extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx:116]: Updated.

File: extension/src/editor/containers/component-template/index.tsx [hunk 3]
```diff
@@ -197,6 +197,43 @@ export const EditorFieldComponentsTemplateSection: React.FC<
...
+        type: "component",
+        groups: template.groups.map(g => ({
+          ...g,
+          id: generateID(),
+          components: g.components.map(c => ({
+            ...c,
+            id: generateID(),
+          })),
+        })),
+      } as unknown as ComponentTemplate;
```
[the remark this answers is not in the document]
Comment (airslice) [reply extension/src/editor/containers/component-template/index.tsx:214]: New ID should be generated on BE, but the type has it

File: extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx [hunk 4]
```diff
@@ -85,6 +82,34 @@ export const TemplateHeader: React.FC<TemplateHeaderProps> = ({
...
+    [],
+  );
+  const duplicatedTemplateNameInvalid = useMemo(() => {
+    return isTemplateNameInvalid(duplicatedTemplateName, templateNames);
+  }, [duplicatedTemplateName, templateNames]);
+
+  const handleTemplateDuplicate = useCallback(() => {
+    onTemplateDuplicate?.(duplicatedTemplateName);
+    setDuplicateTemplateOpen(false);
+  }, [duplicatedTemplateName, onTemplateDuplicate]);
```
[the remark this answers is not in the document]
Comment (airslice) [reply extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx:111]: Already handled the error tracking and notification on higher level hooks.

File: extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx [hunk 5]
```diff
@@ -167,3 +211,16 @@ const StyledButton = styled(Button)(({ theme }) => ({
     borderRadius: "0",
   },
 }));
+
+function generateDuplicateTemplateName(ori: string) {
+  return `${ori} (copy-${new Date().getTime()})`;
```
Comment (keiya01) [inline extension/src/editor/containers/common/commonTemplate/TemplateHeader.tsx:216]: It looks like non-human readable number. Is it no problem? I thought this number confuses user.

**Review by keiya01 — APPROVED** [review]

**Pull request merged**